File: lib/python/opentoken/src/test/opentoken/io/csv/token_csv_reader_test.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import os
import tempfile

import pytest

from opentoken.io.csv.token_csv_reader import TokenCSVReader
from opentoken.processor.token_constants import TokenConstants


class TestTokenCSVReader:
    """Test cases for TokenCSVReader."""

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.temp_file = tempfile.NamedTemporaryFile(mode='w', suffix='.csv', delete=False)
        self.temp_file_path = self.temp_file.name
        self.temp_file.close()
        self.reader = None

    def teardown_method(self):
        """Clean up after each test method."""
        if self.reader:
            self.reader.close()
        if os.path.exists(self.temp_file_path):
            os.unlink(self.temp_file_path)

    def test_read_single_token(self):
        """Test reading a single token from CSV."""
        with open(self.temp_file_path, 'w', encoding='utf-8') as f:
            f.write("RuleId,Token,RecordId\n")
            f.write("T1,abc123token,rec-001\n")

        self.reader = TokenCSVReader(self.temp_file_path)
        
        data = next(self.reader)
        
        assert data[TokenConstants.RULE_ID] == "T1"
        assert data[TokenConstants.TOKEN] == "abc123token"
        assert data[TokenConstants.RECORD_ID] == "rec-001"
        
        with pytest.raises(StopIteration):
            next(self.reader)

    def test_read_multiple_tokens(self):
        """Test reading multiple tokens from CSV."""
        with open(self.temp_file_path, 'w', encoding='utf-8') as f:
            f.write("RuleId,Token,RecordId\n")
            f.write("T1,token1,rec-001\n")
            f.write("T2,token2,rec-002\n")

        self.reader = TokenCSVReader(self.temp_file_path)
        
        data1 = next(self.reader)
        assert data1[TokenConstants.RULE_ID] == "T1"
        assert data1[TokenConstants.TOKEN] == "token1"
        assert data1[TokenConstants.RECORD_ID] == "rec-001"
        
        data2 = next(self.reader)
        assert data2[TokenConstants.RULE_ID] == "T2"
        assert data2[TokenConstants.TOKEN] == "token2"
        assert data2[TokenConstants.RECORD_ID] == "rec-002"
        
        with pytest.raises(StopIteration):
            next(self.reader)

    def test_read_empty_token(self):
        """Test reading a token with empty value."""
        with open(self.temp_file_path, 'w', encoding='utf-8') as f:
            f.write("RuleId,Token,RecordId\n")
            f.write("T1,,rec-001\n")

        self.reader = TokenCSVReader(self.temp_file_path)
        
        data = next(self.reader)
        
        assert data[TokenConstants.RULE_ID] == "T1"
        assert data[TokenConstants.TOKEN] == ""
        assert data[TokenConstants.RECORD_ID] == "rec-001"

    def test_missing_required_column(self):
        """Test that missing required column raises error."""
        with open(self.temp_file_path, 'w', encoding='utf-8') as f:
            f.write("RuleId,Token\n")
            f.write("T1,token1\n")

        with pytest.raises(ValueError, match="Missing required columns"):
            TokenCSVReader(self.temp_file_path)

    def test_empty_file(self):
        """Test that empty file raises error."""
        with open(self.temp_file_path, 'w', encoding='utf-8') as f:
            f.write("")

        with pytest.raises(ValueError, match="Missing required columns"):
            TokenCSVReader(self.temp_file_path)

    def test_iterator_interface(self):
        """Test that reader works as an iterator."""
        with open(self.temp_file_path, 'w', encoding='utf-8') as f:
            f.write("RuleId,Token,RecordId\n")
            f.write("T1,token1,rec-001\n")
            f.write("T2,token2,rec-002\n")

        self.reader = TokenCSVReader(self.temp_file_path)
        
        count = 0
        for data in self.reader:
            count += 1
            assert TokenConstants.RULE_ID in data
            assert TokenConstants.TOKEN in data
            assert TokenConstants.RECORD_ID in data
        
        assert count == 2

File: lib/java/opentoken/src/test/java/com/truveta/opentoken/io/csv/TokenCSVReaderTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.csv;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.truveta.opentoken.processor.TokenConstants;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.util.Map;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class TokenCSVReaderTest {

    private File tempFile;
    private TokenCSVReader reader;

    @BeforeEach
    void setUp() throws Exception {
        tempFile = Files.createTempFile("test_token_", ".csv").toFile();
    }

    @AfterEach
    void tearDown() throws Exception {
        if (reader != null) {
            reader.close();
        }
        if (tempFile.exists()) {
            tempFile.delete();
        }
    }

    @Test
    void testReadSingleToken() throws Exception {
        try (FileWriter writer = new FileWriter(tempFile)) {
            writer.write("RuleId,Token,RecordId\n");
            writer.write("T1,abc123token,rec-001\n");
        }

        reader = new TokenCSVReader(tempFile.getAbsolutePath());
        
        assertTrue(reader.hasNext());
        Map<String, String> data = reader.next();
        
        assertEquals("T1", data.get(TokenConstants.RULE_ID));
        assertEquals("abc123token", data.get(TokenConstants.TOKEN));
        assertEquals("rec-001", data.get(TokenConstants.RECORD_ID));
        
        assertFalse(reader.hasNext());
    }

    @Test
    void testReadMultipleTokens() throws Exception {
        try (FileWriter writer = new FileWriter(tempFile)) {
            writer.write("RuleId,Token,RecordId\n");
            writer.write("T1,token1,rec-001\n");
            writer.write("T2,token2,rec-002\n");
        }

        reader = new TokenCSVReader(tempFile.getAbsolutePath());
        
        assertTrue(reader.hasNext());
        Map<String, String> data1 = reader.next();
        assertEquals("T1", data1.get(TokenConstants.RULE_ID));
        assertEquals("token1", data1.get(TokenConstants.TOKEN));
        assertEquals("rec-001", data1.get(TokenConstants.RECORD_ID));
        
        assertTrue(reader.hasNext());
        Map<String, String> data2 = reader.next();
        assertEquals("T2", data2.get(TokenConstants.RULE_ID));
        assertEquals("token2", data2.get(TokenConstants.TOKEN));
        assertEquals("rec-002", data2.get(TokenConstants.RECORD_ID));
        
        assertFalse(reader.hasNext());
    }

    @Test
    void testReadEmptyToken() throws Exception {
        try (FileWriter writer = new FileWriter(tempFile)) {
            writer.write("RuleId,Token,RecordId\n");
            writer.write("T1,,rec-001\n");
        }

        reader = new TokenCSVReader(tempFile.getAbsolutePath());
        
        assertTrue(reader.hasNext());
        Map<String, String> data = reader.next();
        
        assertEquals("T1", data.get(TokenConstants.RULE_ID));
        assertEquals("", data.get(TokenConstants.TOKEN));
        assertEquals("rec-001", data.get(TokenConstants.RECORD_ID));
        
        assertFalse(reader.hasNext());
    }

    @Test
    void testMissingRequiredColumn() throws Exception {
        try (FileWriter writer = new FileWriter(tempFile)) {
            writer.write("RuleId,Token\n");
            writer.write("T1,token1\n");
        }

        assertThrows(IOException.class, () -> {
            new TokenCSVReader(tempFile.getAbsolutePath());
        });
    }

    @Test
    void testEmptyFile() throws Exception {
        try (FileWriter writer = new FileWriter(tempFile)) {
            writer.write("");
        }

        assertThrows(IOException.class, () -> {
            new TokenCSVReader(tempFile.getAbsolutePath());
        });
    }

    @Test
    void testIteratorInterface() throws Exception {
        try (FileWriter writer = new FileWriter(tempFile)) {
            writer.write("RuleId,Token,RecordId\n");
            writer.write("T1,token1,rec-001\n");
            writer.write("T2,token2,rec-002\n");
        }

        reader = new TokenCSVReader(tempFile.getAbsolutePath());
        
        int count = 0;
        while (reader.hasNext()) {
            Map<String, String> data = reader.next();
            count++;
            assertTrue(data.containsKey(TokenConstants.RULE_ID));
            assertTrue(data.containsKey(TokenConstants.TOKEN));
            assertTrue(data.containsKey(TokenConstants.RECORD_ID));
        }
        
        assertEquals(2, count);
    }
}
